dm-agent/src/main.rs

```diff
@@ -0,0 +1,190 @@
+const DEFAULT_PORT: u16 = 1883;
+const CLIENT_ID: &str = "tedge-dm-agent";
+const SOURCE_TOPIC: &str = "collectd/#";
+const TARGET_TOPIC: &str = "tedge/measurements";
+
+const APP_NAME: &str = "tedge-dm-agent";
+const DEFAULT_LOG_LEVEL: &str = "warn";
+const TIME_FORMAT: &str = "%Y-%m-%dT%H:%M:%S%.3f%:z";
+
+#[tokio::main]
+async fn main() {
+    let filter = std::env::var("RUST_LOG").unwrap_or_else(|_| DEFAULT_LOG_LEVEL.into());
+    tracing_subscriber::fmt()
+        .with_timer(tracing_subscriber::fmt::time::ChronoUtc::with_format(
+            TIME_FORMAT.into(),
+        ))
+        .with_env_filter(filter)
+        .with_span_events(tracing_subscriber::fmt::format::FmtSpan::CLOSE)
+        .init();
+
+    info!("{} starting!", APP_NAME);
+
+    if let Err(err) = run().instrument(debug_span!(APP_NAME)).await {
+        error!("{}", err);
+    }
+}
+
+#[instrument(name = "main")]
+async fn run() -> Result<(), anyhow::Error> {
+    let config = mqtt_client::Config::new(DEFAULT_HOST, DEFAULT_PORT);
+    let mqtt = Arc::new(Client::connect(CLIENT_ID, &config).await?);
+    let filter = TopicFilter::new(SOURCE_TOPIC)?.qos(QoS::AtMostOnce);
+
+    let mut errors = mqtt.subscribe_errors();
+    let mut messages: MessageStream = mqtt.subscribe(filter).await?;
+
+    let mut interval =
+        tokio::time::interval(Duration::from_millis(DEFAULT_STATS_COLLECTION_INTERVAL));
+
+    let mut message_batch = MeasurementGrouper::new();
+
+    loop {
+        select! {
+            error = errors.next().fuse() => {
+                if let Some(err) = error {
+                    error!("{}", err);
+                }
+            }
+
+            maybe_message = messages.next().fuse() => {
+                if let Some(message) = maybe_message {
+                    if message_batch.is_empty() {
+                        let _ = message_batch.timestamp(&current_timestamp()).map_err(|err| error!("{}", err));
+                    }
+                    let _ = tedge_json_from_collectd_message(&mut message_batch, message).map_err(|err| error!("{}", err));
+                }
+            }
+
+            _result = interval.tick() => {
+                if !message_batch.is_empty() {
+                    handle_message_batch(mqtt.clone(), message_batch);
+                    message_batch = MeasurementGrouper::new();
+                }
+            }
+        }
+    }
+}
+
+fn tedge_json_from_collectd_message(
+    message_batch: &mut MeasurementGrouper,
+    message: Message,
+) -> anyhow::Result<()> {
+    let topic = message.topic.name.clone();
+    let split: Vec<&str> = topic.split('/').collect();
+    let group_measurement_key = split[split.len() - 2];
+    let measurement_key = split[split.len() - 1];
+
+    let payload = message.payload;
+    let payload = from_utf8(&payload)?;
+    let split: Vec<&str> = payload.split(':').collect();
+
+    let measurement_value = split[1].trim_end_matches(char::from(0)).parse::<f64>()?;
```

Comment:
Again, check that there is at least one ":" in the payload, otherwise raise error.